sumcheck/src/prover.rs

```diff
@@ -428,110 +428,13 @@ impl<'a, E: ExtensionField> IOPProverState<'a, E> {
-                        )
-                        .to_vec()
-                    }
-                    3 => {
-                        let (f1, f2, f3) = (
-                            &self.poly.flattened_ml_extensions[products[0]],
-                            &self.poly.flattened_ml_extensions[products[1]],
-                            &self.poly.flattened_ml_extensions[products[2]],
-                        );
-                        op_mle_product_3!(
-                            |f1, f2, f3| {
-                                let res = (0..largest_even_below(f1.len()))
-                                    .step_by(2)
-                                    .rev()
-                                    .map(|b| {
-                                        // f = c x + d
-                                        let c1 = f1[b + 1] - f1[b];
-                                        let c2 = f2[b + 1] - f2[b];
-                                        let c3 = f3[b + 1] - f3[b];
-                                        AdditiveArray([
-                                            f1[b] * (f2[b] * f3[b]),
-                                            f1[b + 1] * (f2[b + 1] * f3[b + 1]),
-                                            (c1 + f1[b + 1])
-                                                * ((c2 + f2[b + 1]) * (c3 + f3[b + 1])),
-                                            (c1 + c1 + f1[b + 1])
-                                                * ((c2 + c2 + f2[b + 1]) * (c3 + c3 + f3[b + 1])),
-                                        ])
-                                    })
-                                    .sum::<AdditiveArray<_, 4>>();
-                                let res = if f1.len() == 1 {
-                                    AdditiveArray::<_, 4>([f1[0] * f2[0] * f3[0]; 4])
-                                } else {
-                                    res
-                                };
-                                let num_vars_multiplicity = self.poly.aux_info.max_num_variables - (ceil_log2(f1.len()).max(1) + self.round - 1);
-                                if num_vars_multiplicity > 0 {
-                                    AdditiveArray(res.0.map(|e| e * E::BaseField::from(1 << num_vars_multiplicity)))
-                                } else {
-                                    res
-                                }
-                            },
-                            |sum| AdditiveArray(sum.0.map(E::from))
-                        )
-                        .to_vec()
-                    }
-                    _ => unimplemented!("do not support degree > 3"),
+                    1 => sumcheck_code_gen!(1, false, |i| &f[products[i]]).to_vec(),
+                    2 => sumcheck_code_gen!(2, false, |i| &f[products[i]]).to_vec(),
+                    3 => sumcheck_code_gen!(3, false, |i| &f[products[i]]).to_vec(),
+                    4 => sumcheck_code_gen!(4, false, |i| &f[products[i]]).to_vec(),
```

Comment:
to experiment with product argument in following up task, we will need up to degree 5